### PR TITLE
Bug 1420976 - Support passing reference-policy in import-image command

### PR DESCRIPTION
**Pull request by soltysh**
Fixes https://bugzilla.redhat.com/show_bug.cgi?id=1420976

@mfojtik @miminar ptal

### CODE REVIEW

**Comment by soltysh** [comment]
[test]

**Comment by soltysh** [comment]
@openshift/api-review for the api change, `ReferencePolicy` in `ImageImportSpec`

**Comment by soltysh** [comment]
Forgot about updating completions.

**Comment by smarterclayton** [comment]
API change approved

**Comment by soltysh** [comment]
I've rebased and re-pushed. I can't reproduce that `pkg/api/serialization_test.go` failure locally :/

File: pkg/cmd/cli/cmd/importimage.go [hunk 1]
```diff
@@ -52,6 +52,7 @@ func NewCmdImportImage(fullName string, f *clientcmd.Factory, out, errout io.Wri
 	cmd.Flags().StringVar(&opts.From, "from", "", "A Docker image repository to import images from")
 	cmd.Flags().BoolVar(&opts.Confirm, "confirm", false, "If true, allow the image stream import location to be set or changed")
 	cmd.Flags().BoolVar(&opts.All, "all", false, "If true, import all tags from the provided source on creation or if --from is specified")
+	cmd.Flags().StringVar(&opts.ReferencePolicy, "reference-policy", sourceReferencePolicy, "Allow to request pullthrough for external image when set to 'local'. Defaults to 'source'.")
```
Comment (smarterclayton) [inline pkg/cmd/cli/cmd/importimage.go]: These should be constants, i.e. `Local` and `Source`

File: pkg/image/api/v1/types.go [hunk 2]
```diff
@@ -417,8 +417,10 @@ type ImageImportSpec struct {
 
 	// ImportPolicy is the policy controlling how the image is imported
 	ImportPolicy TagImportPolicy `json:"importPolicy,omitempty" protobuf:"bytes,3,opt,name=importPolicy"`
+	// ReferencePolicy defines how other components should consume the image
+	ReferencePolicy TagReferencePolicy `json:"referencePolicy,omitempty" protobuf:"bytes,4,opt,name=referencePolicy"`
 	// IncludeManifest determines if the manifest for each image is returned in the response
-	IncludeManifest bool `json:"includeManifest,omitempty" protobuf:"varint,4,opt,name=includeManifest"`
+	IncludeManifest bool `json:"includeManifest,omitempty" protobuf:"varint,5,opt,name=includeManifest"`
```
Comment (smarterclayton) [inline pkg/image/api/v1/types.go]: Yo can't do this - you just broke the protobuf serialization.  *Never* change the tag number on a field.

File: pkg/cmd/cli/cmd/importimage.go (hunk 1 above)
Comment (soltysh) [inline pkg/cmd/cli/cmd/importimage.go]: They are in [pkg/cmd/cli/cmd/tag.go](https://github.com/openshift/origin/blob/9e4e2e788bd6d42f2258efe218c71af1c62a64b2/pkg/cmd/cli/cmd/tag.go#L79-L80).

File: pkg/image/api/v1/types.go (hunk 2 above)
Comment (soltysh) [inline pkg/image/api/v1/types.go]: Oooops... my bad. Fixing right away.

**Comment by soltysh** [comment]
Comments addressed.

File: pkg/cmd/cli/cmd/importimage.go [hunk 3]
```diff
@@ -118,6 +123,10 @@ func (o *ImportImageOptions) Validate(cmd *cobra.Command) error {
 		return kcmdutil.UsageError(cmd, "you must specify the name of an image stream")
 	}
 
+	if len(o.ReferencePolicy) > 0 && o.All {
```
Comment (mfojtik) [inline pkg/cmd/cli/cmd/importimage.go]: open an issue about adding support for this in the future

File: pkg/cmd/cli/cmd/importimage.go [hunk 4]
```diff
@@ -440,16 +463,13 @@ func (o *ImportImageOptions) newImageStream() (*imageapi.ImageStream, *imageapi.
 							Kind: "DockerImage",
 							Name: from,
 						},
+						ReferencePolicy: imageapi.TagReferencePolicy{
```
Comment (mfojtik) [inline pkg/cmd/cli/cmd/importimage.go]: i guess this should not be set at all if the reference policy is not set...

i would check below this and generate the ReferencePolicy only if `len(referencePolicy) > 0`

File: pkg/cmd/cli/cmd/importimage.go [hunk 5]
```diff
@@ -498,6 +525,9 @@ func (o *ImportImageOptions) newImageStreamImportTags(stream *imageapi.ImageStre
 			},
 			To:           &kapi.LocalObjectReference{Name: tag},
 			ImportPolicy: imageapi.TagImportPolicy{Insecure: insecure},
+			ReferencePolicy: imageapi.TagReferencePolicy{
```
Comment (mfojtik) [inline pkg/cmd/cli/cmd/importimage.go]: same comment as above

File: pkg/cmd/cli/cmd/importimage_test.go [hunk 6]
```diff
@@ -338,6 +339,28 @@ func TestCreateImageImport(t *testing.T) {
 				ImportPolicy: imageapi.TagImportPolicy{Insecure: false},
 			}},
 		},
+		"import tag setting referencePolicy": {
```
Comment (mfojtik) [inline pkg/cmd/cli/cmd/importimage_test.go]: maybe add test with invalid reference policy? (optional ;)

File: pkg/cmd/cli/cmd/importimage.go (hunk 3 above)
Comment (soltysh) [inline pkg/cmd/cli/cmd/importimage.go]: How would you wanted to work? It's not a problem to support this right away. The thing is that **all** tags will get the same reference policy, including overwriting current ones.

File: pkg/cmd/cli/cmd/importimage.go (hunk 4 above)
Comment (soltysh) [inline pkg/cmd/cli/cmd/importimage.go]: sgtm

File: pkg/cmd/cli/cmd/importimage.go (hunk 5 above)
Comment (soltysh) [inline pkg/cmd/cli/cmd/importimage.go]: k

File: pkg/cmd/cli/cmd/importimage_test.go (hunk 6 above)
Comment (soltysh) [inline pkg/cmd/cli/cmd/importimage_test.go]: sure

Comment (soltysh) [inline pkg/cmd/cli/cmd/importimage_test.go]: On second thought, I don't check `Validate` anywhere, so I'll drop the idea ;

**Comment by soltysh** [comment]
The empty tag reference comments addressed. Rest is either under discussion or will not be fixed.

File: pkg/cmd/cli/cmd/importimage.go (hunk 3 above)
Comment (soltysh) [inline pkg/cmd/cli/cmd/importimage.go]: Sure, that's not a problem. Will update.

**Comment by soltysh** [comment]
@miminar comment addressed.

File: pkg/cmd/cli/cmd/importimage.go (hunk 3 above)
Comment (miminar) [inline pkg/cmd/cli/cmd/importimage.go]: Thanks!

**Comment by soltysh** [comment]
[merge]

**Comment by soltysh** [comment]
Flake https://github.com/openshift/origin/issues/13386

[merge]

**Comment by soltysh** [comment]
Looks like https://github.com/openshift/origin/issues/11016.

[merge]

**Comment by smarterclayton** [comment]
Some examples of 110016 are image pulls - I found that out yesterday with
the router.

We need to start doing image prepulling and simplify the number of images
we use

On Mar 15, 2017, at 2:54 AM, OpenShift Bot <notifications@github.com> wrote:

Merged #13339 <https://github.com/openshift/origin/pull/13339>.

—
You are receiving this because you are on a team that was mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin/pull/13339#event-1000357816>, or mute
the thread

**Comment by soltysh** [comment]
> We need to start doing image prepulling and simplify the number of images we use

When and where?

**Comment by smarterclayton** [comment]
Soon and in the jobs (somehow).

**Pull request merged**